crates/core/src/host/host_controller.rs

```diff
@@ -788,6 +804,61 @@ impl Host {
         })
     }
 
+    /// Construct an in-memory instance of `database` running `program`,
+    /// initialize it, then immediately destroy it.
+    ///
+    /// This is used during an initial, fresh publish operation
+    /// in order to check the `program`'s validity as a module,
+    /// since some validity checks we'd like to do (e.g. typechecking RLS filters)
+    /// require a fully instantiated database.
+    ///
+    /// This is not necessary during hotswap publishes,
+    /// as the automigration planner and executor accomplish the same validity checks.
+    async fn try_init_in_memory_to_check(
+        host_controller: &HostController,
+        database: Database,
+        program: Program,
+    ) -> anyhow::Result<()> {
+        let HostController { runtimes, .. } = host_controller;
+
+        // Even in-memory databases acquire a lockfile.
+        // Grab a tempdir to put that lockfile in.
+        let phony_replica_dir = TempDir::with_prefix("spacetimedb-publish-in-memory-check")
+            .context("Error creating temporary directory to house temporary database during publish")?;
+
+        // Leave the `TempDir` instance in place, so that its destructor will still run.
```

Comment:
And that actually works when the previous binding is no longer reachable? I will never be able to pass rust pop quizzes I guess.

Reply:
Yeah, the object lives until the end of its containing scope, regardless of whether the binding becomes unreachable.